File: python/98_Validate_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.DFS(root, -math.inf, math.inf)
    
    def DFS(self, root: Optional[TreeNode], minVal: int, maxVal: int) -> bool:
        if root.val <= minVal or root.val >= maxVal:
            return False
        if root.left:
            if not self.DFS(root.left, minVal, root.val):
                return False
        if root.right:
            if not self.DFS(root.right, root.val, maxVal):
                return False
        
        return True
            
        
        

File: swift/98_validate_binary_search_tree.swift
/// Definition for a binary tree node.
/// public class TreeNode {
///     public var val: Int
///     public var left: TreeNode?
///     public var right: TreeNode?
///     public init() { self.val = 0; self.left = nil; self.right = nil; }
///     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
///     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
///         self.val = val
///         self.left = left
///         self.right = right
///     }
/// }
class Solution {
  func isValidBST(_ root: TreeNode?) -> Bool {
    return dfs(root, Int.min, Int.max)
  }

  private func dfs(_ node: TreeNode?, _ leftBoundary: Int, _ rightBoundary: Int) -> Bool {
    guard let node = node else { return true }

    if !(leftBoundary < node.val && node.val < rightBoundary) {
      return false
    }

    return dfs(node.left, leftBoundary, node.val) && dfs(node.right, node.val, rightBoundary)
  }
}
